api/inc/secure_access.h

```diff
@@ -86,4 +52,40 @@ static UVISOR_FORCEINLINE uint8_t uvisor_read8(uint8_t volatile * volatile addr)
     return UVISOR_ASM_MEMORY_ACCESS(ldrb, uint8_t, addr);
 }
 
+/* The switch statement will be optimised away since the compiler already knows
+ * the sizeof_type. */
+static UVISOR_FORCEINLINE void __address_write(size_t sizeof_type, volatile uint32_t *addr, uint32_t val)
+{
+    switch(sizeof_type) {
+        case 4:
+            uvisor_write32((volatile uint32_t * volatile) addr, (uint32_t) val);
+            break;
+        case 2:
+            uvisor_write16((volatile uint16_t * volatile) addr, (uint16_t) val);
+            break;
+        case 1:
+            uvisor_write8((volatile uint8_t * volatile) addr, (uint8_t) val);
+            break;
+        default:
+            uvisor_error(USER_NOT_ALLOWED);
+            break;
+    }
+}
+
+#define ADDRESS_WRITE(type, addr, val) __address_write(sizeof(type), (volatile uint32_t *) addr, (uint32_t) val)
+
+/* the conditional statement will be optimised away since the compiler already
+ * knows the sizeof(type) */
+#define ADDRESS_READ(type, addr) \
+    (sizeof(type) == 4 ? uvisor_read32((uint32_t volatile * volatile) (addr)) : \
```

Comment:
@AlessandroA : turn that into an inline with switch statement, too for consistency ?